fix: Introduce type inference using Type.ml

nmote opened the pull request:
Here, I bring `Type.ml` over from the Pro Engine, and use it to it to implement a couple cases of type inference, just enough to showcase the idea. I don't think that `AST_generic.type_` is suitable for use as a type representation for any sophisticated type inference.

This way, we can build out local type inference in OSS, improving matching for typed metavariables and possibly other things when local type inference is able to determine a type. We can also make use of the Pro Engine's resolved types for symbols defined outside of the local scope to further improve typed metavariable matching in the Pro Engine.

Test plan: Automated tests

PR checklist:

- [x] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [x] Tests included or PR comment includes a reproducible test plan
- [x] Documentation is up-to-date
- [x] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [x] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

aryx commented on src/core/Type.ml:
```diff
@@ -0,0 +1,211 @@
+(*****************************************************************************)
+type todo_kind = string option [@@deriving show]
+
+(* Fully qualified name *)
```
maybe explain why we use a polymorphic type because this module is also used in semgrep-pro
where the resolved name is difference.

aryx replied:
Also maybe time to define a Type.mli given the non-type-definition parts is getting big and .mli are better
documentation in that case.

nmote replied:
I'm not a fan of repeating potentially-large variant types in an `mli`. Especially because in this case I anticipate generating visitors for it at some point, which aren't compatible with an `mli`.

nmote replied:
Oh wait, I already wrote a comment that explains the polymorphism.

aryx commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -1436,8 +1436,11 @@ and m_compatible_type lang typed_mvar t e =
  * as well as an ident option that can then be used to query LSP to get the
  * type of the ident.
+ *
+ * Old type inference over `AST_generic.type_`. Let's migrate all of this to the
```
add a TODO:

aryx commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -1499,6 +1502,47 @@ and type_of_expr lang e : G.type_ option * G.ident option =
+ * New Type inference over `Type.t`. Prefer this to the old type inference over
+ * `AST_generic.type_`. Eventually we'll do all the type inference here and
+ * delete the old. *)
+and type_of_expr_new lang e : G.name Type.t * G.ident option =
```
Could we move this code out of Generic_vs_generic.ml? in a typing/Typing.ml ?

nmote replied:
Sure -- I was contemplating something like this anyway. Thanks for the nudge.

aryx commented on src/matching/Generic_vs_generic.ml:
```diff
@@ -1499,6 +1502,47 @@ and type_of_expr lang e : G.type_ option * G.ident option =
+  | _else_ -> (Type.NoType, None)
+
+and type_of_expr lang e : G.type_ option * G.ident option =
+  let t, id = type_of_expr_new lang e in
```
but this time 't' contains only fake tokens right? Is it an issue? You remember
the mess there was with the Name.t refactoring in semgrep-pro, where resolved names
suddently were "pure" without tokens but then we had to pass the token of an existing
id when converting back to the generic AST.

nmote replied:
Correct. I think this is going to happen if we do type inference anyway, even if we use `AST_generic.type_`. We'll end up pulling types out of thin air in that case too, so we'll need to deal with having fake tokens.

I don't think it matters in this case though. We use this type for typed metavariable matching, but the metavariable still binds to the original expression, not the inferred type. The issues that I recall with fake tokens in synthetic AST nodes were related to metavariable bindings. I think this type should just be used to determine whether or not to match.

nmote committed:
fix: Introduce type inference using Type.ml

Here, I bring `Type.ml` over from the Pro Engine, and use it to it to
implement a couple cases of type inference, just enough to showcase the
idea. I don't think that `AST_generic.type_` is suitable for use as a
type representation for any sophisticated type inference.

This way, we can build out local type inference in OSS, improving
matching for typed metavariables and possibly other things when local
type inference is able to determine a type. We can also make use of the
Pro Engine's resolved types for symbols defined outside of the local
scope to further improve typed metavariable matching in the Pro Engine.

Test plan: Automated tests
